File: sanitized/thought-vault/app/src/main/java/com/example/thoughtvault/data/remote/WebdavApi.kt
package com.example.thoughtvault.data.remote

import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import javax.inject.Inject
import javax.inject.Singleton
import timber.log.Timber

/**
 * WebDAV 高级 API — 定义 URL 构造和文件操作的上层语义。
 */
@Singleton
class WebdavApi @Inject constructor(
    private val client: WebdavClient
) {
    companion object {
        private val DATE_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd")
        private val TIME_FMT = DateTimeFormatter.ofPattern("HH:mm")
    }

    /** 构建完整的 WebDAV URL */
    fun buildUrl(baseUrl: String, path: String): String {
        val base = baseUrl.trimEnd('/')
        val p = path.trimStart('/')
        return "$base/$p"
    }

    /** 原始记录文件路径: raw/YYYY/MM/YYYY-MM-DD.md */
    fun rawFilePath(date: LocalDate): String {
        val year = date.year
        val month = String.format("%02d", date.monthValue)
        val day = date.format(DATE_FMT)
        return "raw/$year/$month/$day.md"
    }

    /** 原始记录文件所在目录路径: raw/YYYY/MM/ */
    fun rawDirPath(date: LocalDate): String {
        val year = date.year
        val month = String.format("%02d", date.monthValue)
        return "raw/$year/$month/"
    }

    /** 锁文件路径 */
    fun lockFilePath(): String = ".sync/write.lock"

    /** 日报文件路径: daily/YYYY/MM/YYYY-MM-DD.md */
    fun dailyFilePath(date: LocalDate): String {
        val year = date.year
        val month = String.format("%02d", date.monthValue)
        return "daily/$year/$month/${date.format(DATE_FMT)}.md"
    }

    /**
     * 递归创建目录路径。WebDAV MKCOL 一次只能创建一级，
     * 所以从根开始逐级确保每一级目录都存在。
     */
    private suspend fun mkcolRecursive(
        baseUrl: String,
        username: String,
        password: String,
        fullPath: String,
    ) {
        // 去掉首尾斜杠，按 / 拆分为各级
        val parts = fullPath.trim('/').split("/")
        var current = ""
        for (part in parts) {
            current = if (current.isEmpty()) part else "$current/$part"
            val dirUrl = buildUrl(baseUrl, current)
            client.mkcol(dirUrl, username, password)
        }
    }

    /**
     * 保存一条新条目（完整的读-改-写 + 锁流程）。
     */
    suspend fun saveEntry(
        baseUrl: String,
        username: String,
        password: String,
        date: LocalDate,
        time: LocalTime,
        content: String,
    ): Result<Unit> {
        val lockUrl = buildUrl(baseUrl, lockFilePath())
        val fileUrl = buildUrl(baseUrl, rawFilePath(date))

        // 1. 确保 .sync 目录存在（WebDAV PUT 要求父目录已存在）
        mkcolRecursive(baseUrl, username, password, ".sync")

        // 2. 获取锁（指数退避重试，最多 10 秒）
        //    如果 3 次后仍失败，检查是否为僵尸锁（>60 秒未释放的锁强制清理）
        var lockAcquired = false
        for (attempt in 1..5) {
            val lockResult = client.putEmpty(lockUrl, username, password)
            if (lockResult.isSuccess) {
                lockAcquired = true
                break
            }
            if (attempt == 3) {
                // 第 3 次失败 → 可能是僵尸锁，强制清理
                Timber.w("锁获取失败 3 次，尝试强制清理僵尸锁...")
                client.delete(lockUrl, username, password)
            }
            if (attempt < 5) {
                val delayMs = 250L * (1L shl (attempt - 1))  // 250, 500, 1000, 2000
                kotlinx.coroutines.delay(delayMs)
            }
        }
        if (!lockAcquired) {
            return Result.failure(Exception("无法获取写入锁，请稍后重试"))
        }

        try {
            // 3. 递归创建目录（WebDAV MKCOL 一次只能建一级）
            mkcolRecursive(baseUrl, username, password, rawDirPath(date))

            // 4. 读取当日已有内容
            val getResult = client.get(fileUrl, username, password)
            if (getResult.isFailure) {
                return Result.failure(getResult.exceptionOrNull()!!)
            }
            val existing = getResult.getOrNull()

            // 4. 构建新内容
            val newContent = if (existing != null) {
                // 追加新条目
                val entryBlock = buildEntryBlock(time, content)
                existing.trimEnd() + "\n\n" + entryBlock
            } else {
                // 新的一天
                val header = "# ${date.format(DATE_FMT)} 原始记录\n\n---\n"
                header + buildEntryBlock(time, content)
            }

            // 5. 写回
            val putResult = client.put(fileUrl, newContent, username, password)
            if (putResult.isFailure) {
                return Result.failure(putResult.exceptionOrNull()!!)
            }

            return Result.success(Unit)
        } finally {
            // 6. 释放锁
            client.delete(lockUrl, username, password)
        }
    }

    /**
     * 读取某天的原始记录。
     */
    suspend fun getRawEntries(
        baseUrl: String,
        username: String,
        password: String,
        date: LocalDate,
    ): Result<String?> {
        val fileUrl = buildUrl(baseUrl, rawFilePath(date))
        return client.get(fileUrl, username, password)
    }

    /**
     * 读取 AI 处理后的日报。
     */
    suspend fun getDailyReport(
        baseUrl: String,
        username: String,
        password: String,
        date: LocalDate,
    ): Result<String?> {
        val fileUrl = buildUrl(baseUrl, dailyFilePath(date))
        return client.get(fileUrl, username, password)
    }

    /**
     * 列出某月所有有记录的日期。
     */
    suspend fun listMonthDays(
        baseUrl: String,
        username: String,
        password: String,
        year: Int,
        month: Int,
    ): Result<String> {
        val path = "raw/$year/${String.format("%02d", month)}/"
        val url = buildUrl(baseUrl, path)
        return client.propfind(url, username, password, depth = "1")
    }

    /** 读取待办文件 */
    suspend fun getTodosFile(
        baseUrl: String, username: String, password: String,
    ): Result<String?> {
        val url = buildUrl(baseUrl, "todos/tasks.md")
        return client.get(url, username, password)
    }

    /** 汇总目录 */
    fun summaryDirPath(type: String): String = "$type/"

    /** 周报文件路径 */
    fun weeklyFilePath(week: String): String {
        val y = week.substring(0, 4)
        return "weekly/$y/$week.md"
    }

    /** 月报文件路径 */
    fun monthlyFilePath(month: String): String {
        val y = month.substring(0, 4)
        return "monthly/$y/$month.md"
    }

    /** 季报文件路径 */
    fun quarterlyFilePath(q: String): String {
        val y = q.substring(0, 4)
        return "quarterly/$y/$q.md"
    }

    /** 年报文件路径 */
    fun yearlyFilePath(year: String): String = "yearly/$year/$year.md"

    /** PROPFIND 列出周报 */
    suspend fun listWeeklyFiles(
        baseUrl: String, username: String, password: String,
    ): Result<String> {
        val url = buildUrl(baseUrl, "weekly/")
        return client.propfind(url, username, password, depth = "infinity")
    }

    /** PROPFIND 列出月报 */
    suspend fun listMonthlyFiles(
        baseUrl: String, username: String, password: String,
    ): Result<String> {
        val url = buildUrl(baseUrl, "monthly/")
        return client.propfind(url, username, password, depth = "infinity")
    }

    /** PROPFIND 列出季报 */
    suspend fun listQuarterlyFiles(
        baseUrl: String, username: String, password: String,
    ): Result<String> {
        val url = buildUrl(baseUrl, "quarterly/")
        return client.propfind(url, username, password, depth = "infinity")
    }

    /** PROPFIND 列出年报 */
    suspend fun listYearlyFiles(
        baseUrl: String, username: String, password: String,
    ): Result<String> {
        val url = buildUrl(baseUrl, "yearly/")
        return client.propfind(url, username, password, depth = "infinity")
    }

    /** GET 汇总文件内容 */
    suspend fun getSummaryFile(
        baseUrl: String, username: String, password: String, path: String,
    ): Result<String?> {
        val url = buildUrl(baseUrl, path)
        return client.get(url, username, password)
    }

    /** 写入待办文件 */
    suspend fun putTodosFile(
        baseUrl: String, username: String, password: String, content: String,
    ): Result<Unit> {
        val url = buildUrl(baseUrl, "todos/tasks.md")
        // 确保目录存在
        mkcolRecursive(baseUrl, username, password, "todos")
        return client.put(url, content, username, password)
    }

    /** 写入原始记录文件（全量覆盖） */
    suspend fun putRawFile(
        baseUrl: String, username: String, password: String,
        date: LocalDate, content: String,
    ): Result<Unit> {
        val fileUrl = buildUrl(baseUrl, rawFilePath(date))
        mkcolRecursive(baseUrl, username, password, rawDirPath(date))
        return client.put(fileUrl, content, username, password)
    }

    /** 写入日报文件（全量覆盖） */
    suspend fun putDailyReport(
        baseUrl: String, username: String, password: String,
        date: LocalDate, content: String,
    ): Result<Unit> {
        val fileUrl = buildUrl(baseUrl, dailyFilePath(date))
        // 确保目录存在
        val year = date.year
        val month = String.format("%02d", date.monthValue)
        mkcolRecursive(baseUrl, username, password, "daily/$year/$month")
        return client.put(fileUrl, content, username, password)
    }

    /**
     * 测试连接。
     */
    suspend fun testConnection(
        baseUrl: String,
        username: String,
        password: String,
    ): Result<Boolean> {
        return client.testConnection(baseUrl, username, password)
    }

    /** 构建一条条目的 Markdown 块 */
    private fun buildEntryBlock(time: LocalTime, content: String): String {
        val timeStr = time.format(TIME_FMT)
        return "## $timeStr\n\n$content"
    }
}

File: thought-vault-web/src/lib/webdav-api.ts
import { WebdavClient } from './webdav-client'
import { type Result, ok, err } from './types'

/**
 * WebDAV 高级 API — 对标 Android WebdavApi.kt
 * 路径前缀包含用户名，支持多用户数据隔离
 */
export class WebdavApi {
  private client: WebdavClient
  private userPrefix: string

  constructor(client: WebdavClient, username: string) {
    this.client = client
    this.userPrefix = `homes/${username}/thoughts/`
  }


  // ── 路径构造（带用户前缀） ──

  rawFilePath(date: string): string {
    const [y, m] = date.split('-')
    return `${this.userPrefix}raw/${y}/${m}/${date}.md`
  }

  rawDirPath(date: string): string {
    const [y, m] = date.split('-')
    return `${this.userPrefix}raw/${y}/${m}/`
  }

  dailyFilePath(date: string): string {
    const [y, m] = date.split('-')
    return `${this.userPrefix}daily/${y}/${m}/${date}.md`
  }

  dailyDirPath(date: string): string {
    const [y, m] = date.split('-')
    return `${this.userPrefix}daily/${y}/${m}/`
  }

  lockFilePath(): string {
    return `${this.userPrefix}.sync/write.lock`
  }

  todosFilePath(): string {
    return `${this.userPrefix}todos/tasks.md`
  }

  /** WebDAV 根路径（用于 testConnection） */
  rootPath(): string {
    return this.userPrefix
  }

  // ── 汇总文件路径 ──

  weeklyFilePath(week: string): string {
    const y = week.slice(0, 4)
    return `${this.userPrefix}weekly/${y}/${week}.md`
  }

  monthlyFilePath(month: string): string {
    const y = month.slice(0, 4)
    return `${this.userPrefix}monthly/${y}/${month}.md`
  }

  quarterlyFilePath(q: string): string {
    const y = q.slice(0, 4)
    return `${this.userPrefix}quarterly/${y}/${q}.md`
  }

  yearlyFilePath(year: string): string {
    return `${this.userPrefix}yearly/${year}/${year}.md`
  }

  /** PROPFIND 列出所有周报 */
  async listWeeklyFiles(): Promise<Result<string>> {
    return this.client.propfind(`${this.userPrefix}weekly/`, 'infinity')
  }

  /** PROPFIND 列出所有月报 */
  async listMonthlyFiles(): Promise<Result<string>> {
    return this.client.propfind(`${this.userPrefix}monthly/`, 'infinity')
  }

  /** PROPFIND 列出所有季报 */
  async listQuarterlyFiles(): Promise<Result<string>> {
    return this.client.propfind(`${this.userPrefix}quarterly/`, 'infinity')
  }

  /** PROPFIND 列出所有年报 */
  async listYearlyFiles(): Promise<Result<string>> {
    return this.client.propfind(`${this.userPrefix}yearly/`, 'infinity')
  }

  /** GET 汇总文件内容 */
  async getSummaryContent(path: string): Promise<Result<string | null>> {
    return this.client.get(path)
  }

  // ── 读操作 ──

  async getRawEntries(date: string): Promise<Result<string | null>> {
    return this.client.get(this.rawFilePath(date))
  }

  async getDailyReport(date: string): Promise<Result<string | null>> {
    return this.client.get(this.dailyFilePath(date))
  }

  async getTodosFile(): Promise<Result<string | null>> {
    return this.client.get(this.todosFilePath())
  }

  /** PROPFIND 列出某月所有有记录的日期 */
  async listMonthDays(year: number, month: number): Promise<Result<string>> {
    const path = `${this.userPrefix}raw/${year}/${String(month).padStart(2, '0')}/`
    return this.client.propfind(path, 1)
  }

  // ── 写操作 ──

  /**
   * 保存一条新条目（完整的读-改-写 + 锁流程）
   */
  async saveEntry(date: string, time: string, content: string): Promise<Result<void>> {
    const lockPath = this.lockFilePath()
    const filePath = this.rawFilePath(date)
    const dirPath = this.rawDirPath(date)

    // 1. 获取写入锁（最多 5 次重试）
    let lockAcquired = false
    for (let attempt = 1; attempt <= 5; attempt++) {
      const lockResult = await this.client.putEmpty(lockPath)
      if (lockResult.success) {
        lockAcquired = true
        break
      }
      if (attempt === 3) {
        console.warn('锁获取失败 3 次，尝试强制清理僵尸锁...')
        await this.client.delete(lockPath)
      }
      if (attempt < 5) {
        const delayMs = 250 * (1 << (attempt - 1))
        await new Promise(r => setTimeout(r, delayMs))
      }
    }
    if (!lockAcquired) {
      return err('无法获取写入锁，请稍后重试')
    }

    try {
      // 2. 确保目录存在
      await this.client.mkcol(dirPath)

      // 3. 读取已有内容
      const getResult = await this.client.get(filePath)
      if (!getResult.success) return err(getResult.error || '读取文件失败')

      const existing = getResult.data || ''

      // 4. 构建新内容（追加新条目）
      const entryBlock = `## ${time}\n\n${content}`
      let newContent: string
      if (existing) {
        newContent = existing.trimEnd() + '\n\n' + entryBlock
      } else {
        newContent = `# ${date} 原始记录\n\n---\n${entryBlock}`
      }

      // 5. 写回
      const putResult = await this.client.put(filePath, newContent)
      if (!putResult.success) return putResult

      return ok(undefined)
    } finally {
      // 6. 释放锁
      await this.client.delete(lockPath)
    }
  }

  /** 全量覆盖原始记录文件 */
  async putRawFile(date: string, content: string): Promise<Result<void>> {
    const dirPath = this.rawDirPath(date)
    await this.client.mkcol(dirPath)
    return this.client.put(this.rawFilePath(date), content)
  }

  /** 全量覆盖日报文件 */
  async putDailyReport(date: string, content: string): Promise<Result<void>> {
    const dirPath = this.dailyDirPath(date)
    await this.client.mkcol(dirPath)
    return this.client.put(this.dailyFilePath(date), content)
  }

  /** 保存待办文件 */
  async putTodosFile(content: string): Promise<Result<void>> {
    await this.client.mkcol(`${this.userPrefix}todos`)
    return this.client.put(this.todosFilePath(), content)
  }

  // ── 连接测试 ──

  async testConnection(): Promise<Result<boolean>> {
    const r = await this.client.propfind(this.rootPath(), 0)
    if (r.success) return { success: true, data: true }
    console.error('testConnection error:', r.error)
    return { success: false, error: r.error }
  }
}
